### DISCUSSION TITLE
Pest v2

### DISCUSSION THREAD

**Discussion by davideprevosto** [dingo-d/wp-pest#26] [General]
Hello,

Thank you for sharing this package.

Do you plan to upgrade it and using Pest V2?

Regards

**Comment by dingo-d** [comment]
Hi!

I was planning to, but the pest v2 requirement is PHP 8.1, so if I'd release v2 it would be limited to PHP 8.1.

I'd have to check out if it would work with my dependencies as intended. Plus WP core itself is not fully compatible with PHP 8.1 yet, so I'm not sure if there would be some underlying issues when running tests.

I'll have to invest some time in testing it out.

**Reply by davideprevosto** [reply]
If you'd like a beta tester or any help... here I am.

Thank you

**Reply by dingo-d** [reply]
@davideprevosto I've released v2 alpha version so you can give it a go and report any issues you've found 👍🏼

**Reply by davideprevosto** [reply]
@dingo-d I am trying to install and use your package.

I run:

`vendor/bin/wp-pest setup --help` 

...and got:

`project-type Select whether you want to setup tests for theme or a plugin. Can be "theme" or "plugin"`

I understood I can use "theme" or "plugin". I am trying to use your package in one of our plugin, so I run:

`vendor/bin/wp-pest setup plugin`

I am getting:

` [ERROR] Plugin slug must be a string.`

Then I understood (I suppose) I have to use:

`vendor/bin/wp-pest setup plugin --plugin-slug the-plugin-slug`

**Reply by dingo-d** [reply]
Yeah, it's mentioned in the readme when you run `vendor/bin/wp-pest setup --help` you'll get

```bash
Description:
  Sets up the test suites.

Usage:
  setup [options] [--] <project-type>

Arguments:
  project-type                     Select whether you want to setup tests for theme or a plugin. Can be "theme" or "plugin"

Options:
      --wp-version[=WP-VERSION]    Pass the version of the WordPress you want to test on. [default: "latest"]
      --plugin-slug[=PLUGIN-SLUG]  If you are setting the plugin tests provide the plugin slug.
      --skip-delete                If you are running the setup tests in a CI pipeline, provide this option to skip the deletion step.
  -h, --help                       Display help for the given command. When no command is given display help for the list command
  -q, --quiet                      Do not output any message
  -V, --version                    Display this application version
      --ansi|--no-ansi             Force (or disable --no-ansi) ANSI output
  -n, --no-interaction             Do not ask any interactive question
  -v|vv|vvv, --verbose             Increase the verbosity of messages: 1 for normal output, 2 for more verbose output and 3 for debug

Help:
  This command helps you set up WordPress integration and unit test suites.
```

I could add which option is required in the help command.

**Comment by davideprevosto** [comment]
Hello,

I am trying to test one of our custom plugin, using the feature/v2 branch.

The composer.json of our plugin is:

```
{
    "name": "encodia/name-plugin",
    "type": "wordpress-plugin",
    "license": "proprietary",
    "description": "Description...",
    "authors": [
        {
            "name": "Encodia",
            "email": "info@encodia.it",
            "homepage": "https://www.encodia.it"
        }
    ],
    "autoload": {
        "psr-4": {
            "Encodia\\DigitalQuotation\\": "src/"
        }
    },
    "autoload-dev": {
        "psr-4": {
            "Tests\\\\": "tests/"
        }
    },
    "minimum-stability": "dev",
    "require": {
        "php": "^8.1",
        "composer/installers": "~2.0",
        "mikehaertl/php-pdftk": "^0.13.0"
    },
    "config": {
        "allow-plugins": {
            "composer/installers": true,
            "pestphp/pest-plugin": true
        }
    },
    "require-dev": {
        "laravel/pint": "^1.10",
        "dingo-d/wp-pest-integration-test-setup": "dev-main"
    },
    "scripts": {
        "lint": "pint",
        "lint:test": "pint --test",
        "test": "vendor/bin/pest"
    }
}
```

Executing vendor/bin/wp-pest setup plugin --plugin-slug the-plugin-slug I am getting this structure in my plugin project: 

<img width="278" alt="Screenshot 2023-09-23 alle 12 49 59" src="https://github.com/dingo-d/wp-pest/assets/28837345/01283443-aa21-4199-abf4-53d29ca2ac1b">

Executing vendor/bin/pest I am getting:

```
❯ vendor/bin/pest
   Pest\Exceptions\TestCaseClassOrTraitNotFound
   The class `Yoast\WPTestUtils\WPIntegration\TestCase` was not found.
```

Can you put me on the right track? I am missing some basic concepts.

Thank you

**Comment by dingo-d** [comment]
Hi @davideprevosto, as mentioned in the readme, to run unit tests, you'll need to run:

```bash
vendor/bin/pest --group=unit
```

And in order to run integration tests, you'll need to run

```bash
vendor/bin/pest --group=integration
```

Be careful about the integration tests, as you'll probably have to do [some modifications](https://github.com/dingo-d/wp-pest#running-unit-tests-alongside-integration-tests) to the test suite in order to run them.

There was also [a comment](https://github.com/dingo-d/wp-pest/issues/25#issuecomment-1591162431) about running integration tests that might help.

**Comment by davideprevosto** [comment]
I have got the same error even executing:

```
$ vendor/bin/pest --group=unit

Pest\Exceptions\TestCaseClassOrTraitNotFound
The class `Yoast\WPTestUtils\WPIntegration\TestCase` was not found.
```

I read the comment and I changed my phpunit.xml as described (can't imagine it, my fault)

```
<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
		 xsi:noNamespaceSchemaLocation="./vendor/phpunit/phpunit/phpunit.xsd"
		 bootstrap="vendor/autoload.php"
		 colors="true"
>
```

...but I am still getting the same error

**Comment by dingo-d** [comment]
I'll have to investigate this a bit more I didn't get errors about no test case being found when running it 🤔

**Reply by davideprevosto** [reply]
@dingo-d maybe I am doing something wrong and you're going to waste your time investigating. Sorry about that.
Let me know if I can be helpful in any way.

**Reply by dingo-d** [reply]
Nah, I need to make sure I didn't miss anything.

Can you share a bit more info about your setup? Like PHP version, Composer version, OS? Maybe give a minimal example so that I can try to recreate the issue.

**Reply by davideprevosto** [reply]
MacOS 13.6 (22G120)
PHP 8.1.x
Composer version 2.6.3 2023-09-15 09:38:21

composer.json
```
{
    "name": "encodia/name-plugin",
    "type": "wordpress-plugin",
    "license": "proprietary",
    "description": "Description...",
    "authors": [
        {
            "name": "Encodia",
            "email": "info@encodia.it",
            "homepage": "https://www.encodia.it"
        }
    ],
    "autoload": {
        "psr-4": {
            "Encodia\\DigitalQuotation\\": "src/"
        }
    },
    "autoload-dev": {
        "psr-4": {
            "Tests\\\\": "tests/"
        }
    },
    "minimum-stability": "dev",
    "require": {
        "php": "^8.1",
        "composer/installers": "~2.0",
        "mikehaertl/php-pdftk": "^0.13.0"
    },
    "config": {
        "allow-plugins": {
            "composer/installers": true,
            "pestphp/pest-plugin": true
        }
    },
    "require-dev": {
        "laravel/pint": "^1.10",
        "dingo-d/wp-pest-integration-test-setup": "dev-main"
    },
    "scripts": {
        "lint": "pint",
        "lint:test": "pint --test",
        "test": "vendor/bin/pest"
    }
}
```

phpunit.xml
```
<?xml version="1.0" encoding="UTF-8"?>
	<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
			 xsi:noNamespaceSchemaLocation="./vendor/phpunit/phpunit/phpunit.xsd"
			 bootstrap="vendor/autoload.php"
			 colors="true"
	>
	<testsuites>
		<testsuite name="Unit Test Suite">
			<directory>./tests/Unit/</directory>
		</testsuite>
		<testsuite name="Integration Test Suite">
			<directory>./tests/Integration/</directory>
		</testsuite>
	</testsuites>
	<php>
		<env name="TEST" value="true" force="true" />
		<env name="WP_TESTS_DIR" value="wp/tests/phpunit"/>
		<server name="DB_NAME" value="wp_pest_test_db"/>
		<server name="DB_USER" value="root"/>
		<server name="DB_PASSWORD" value=""/>
		<server name="DB_HOST" value="localhost"/>
	</php>
	<coverage processUncoveredFiles="true">
		<include>
			<directory suffix=".php">./src</directory>
		</include>
		<report>
			<clover outputFile="tests/coverage/clover.xml"/>
			<html outputDirectory="tests/coverage/html" lowUpperBound="50" highLowerBound="90"/>
		</report>
	</coverage>
</phpunit>
```
```

vendor/bin/pest --group=unit
Pest\Exceptions\TestCaseClassOrTraitNotFound
The class `Yoast\WPTestUtils\WPIntegration\TestCase` was not found.
```

**Comment by dingo-d** [comment]
Since you put the `bootstrap="vendor/autoload.php"` in your `phpunit.xml` you should add the 

```php
require 'bootstrap.php';
```

before the `use` statement in `Pest.php` file.

I just tried it on my personal project, and I got it to work (both unit and integration tests).

I have these scripts in my `composer.json`

```json
	...
	"scripts": {
		"test:unit": "@php ./vendor/bin/pest --group=unit",
		"test:integration": "@php ./vendor/bin/pest --group=integration"
	},
```

I updated the config to this:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:noNamespaceSchemaLocation="./vendor/phpunit/phpunit/phpunit.xsd"
         bootstrap="vendor/autoload.php"
         colors="true"
>
	<testsuites>
		<testsuite name="Unit Test Suite">
			<directory>./tests/Unit/</directory>
		</testsuite>
		<testsuite name="Integration Test Suite">
			<directory>./tests/Integration/</directory>
		</testsuite>
	</testsuites>
	<php>
		<env name="WP_TESTS_DIR" value="wp/tests/phpunit"/>
		<env name="TEST" value="true" force="true"/>
		<server name="DB_NAME" value="wp_pest_test_db"/>
		<server name="DB_USER" value="root"/>
		<server name="DB_PASSWORD" value=""/>
		<server name="DB_HOST" value="localhost"/>
	</php>
	<source>
		<include>
			<directory suffix=".php">./src</directory>
		</include>
	</source>
	<coverage>
		<report>
			<clover outputFile="tests/coverage/clover.xml"/>
			<html outputDirectory="tests/coverage/html" lowUpperBound="50" highLowerBound="90"/>
		</report>
	</coverage>
</phpunit>
```

**Reply by davideprevosto** [reply]
I started from scratch, following this link https://github.com/dingo-d/wp-pest/tree/feature/v2

I made this steps:

1. In the composer.json file of my plugin I added: `"minimum-stability": "dev",`
2. In the require-dev section of my composer.json file I added:
```
"require-dev": {
   "dingo-d/wp-pest-integration-test-setup": "dev-feature/v2"
},
```
3. I run composer update and then I executed: 
```shell
vendor/bin/wp-pest setup plugin --plugin-slug digital-quotation # (digital-quotation is my plugin slug)
``` 

4. I executed: 
```shell
vendor/bin/pest --group=unit
``` 
...and I've got this warning:

```shell
Your XML configuration validates against a deprecated schema. Migrate your XML configuration using "--migrate-configuration"!
```

5. I executed: 
```shell
vendor/bin/pest --migrate-configuration 
vendor/bin/pest --group=unit
```

The Unit test has been passed!

<img width="469" alt="image" src="https://github.com/dingo-d/wp-pest/assets/28837345/450f5170-1503-481e-b8cc-3f6bcdc718f8">

6. I run 
```shell
vendor/bin/pest --group=integration
``` 
...and I've got an error:

```
Warning: require_once(/Users/encodia/Sviluppo/www/customer/digital-quotation/wp/tests/phpunit/includes/functions.php): Failed to open stream: No such file or directory in /Users/encodia/Sviluppo/www/customer/digital-quotation/tests/bootstrap.php on line 41
```
At boostrap.php:41 I can see: 

```php
require_once dirname(__FILE__, 2) . '/wp/tests/phpunit/includes/functions.php';
```
<img width="400" alt="image" src="https://github.com/dingo-d/wp-pest/assets/28837345/77cd3011-c652-43e9-ab38-0dc9254e7f3e">

There isn't a "includes" folder and I am still investigating.

For references:

phpunit.xml

```xml
<?xml version="1.0" encoding="UTF-8"?>
<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/10.3/phpunit.xsd" bootstrap="tests/bootstrap.php" colors="true" cacheDirectory=".phpunit.cache">
  <testsuites>
    <testsuite name="Unit Test Suite">
      <directory>./tests/Unit/</directory>
    </testsuite>
    <testsuite name="Integration Test Suite">
      <directory>./tests/Integration/</directory>
    </testsuite>
  </testsuites>
  <php>
    <env name="TEST" value="true" force="true"/>
    <server name="DB_NAME" value="wp_pest_test_db"/>
    <server name="DB_USER" value="root"/>
    <server name="DB_PASSWORD" value=""/>
    <server name="DB_HOST" value="localhost"/>
  </php>
  <coverage>
    <report>
      <clover outputFile="tests/coverage/clover.xml"/>
      <html outputDirectory="tests/coverage/html" lowUpperBound="50" highLowerBound="90"/>
    </report>
  </coverage>
  <source>
    <include>
      <directory suffix=".php">./src</directory>
    </include>
  </source>
</phpunit>
```

composer.json

```json
{
    "name": "encodia/digital-quotation",
    "type": "wordpress-plugin",
    "license": "Proprietary",
    "description": "Crea...",
    "authors": [
        {
            "name": "Encodia",
            "email": "info@encodia.it",
            "homepage": "https://www.encodia.it"
        }
    ],
    "autoload": {
        "psr-4": {
            "Encodia\\DigitalQuotation\\": "src/"
        }
    },
    "minimum-stability": "dev",
    "require": {
        "php": ">=8.0.0",
        "composer/installers": "^2.2",
        "mikehaertl/php-pdftk": "^0.13.0"
    },
    "require-dev": {
        "dingo-d/wp-pest-integration-test-setup": "dev-feature/v2"
    },
    "config": {
        "allow-plugins": {
            "composer/installers": true,
            "pestphp/pest-plugin": true
        }
    },
    "scripts": {
        "format": "pint",
        "format:test": "pint --test"
    }
}
```

Pest.php

```php
<?php

require 'bootstrap.php';

use Yoast\WPTestUtils\BrainMonkey\TestCase;

uses()->group('integration')->in('Integration');
uses()->group('unit')->in('Unit');

uses(TestCase::class)->in('Unit', 'Integration');
```

**Reply by davideprevosto** [reply]
I suppose there's something related the the wp-content directory creation. Sometime is working, sometimes not. Not yet understand, but in such cases the db.php was not copied to the wp/src/wp-content folder, because ./wp-content folder wasn't there.

After restarting from scratch several times, I am now getting:

```shell
vendor/bin/pest --group=integration

INFO  Installing...

Running as single site... To run multisite, use -c tests/phpunit/multisite.xml

Warning: Class "PHPUnit\Framework\Error\Deprecated" not found in /Users/encodia/Sviluppo/www/customer/digital-quotation/wp/tests/phpunit/includes/phpunit6/compat.php on line 8
```

**Reply by dingo-d** [reply]
In your `phpunit.xml` file replace `bootstrap="tests/bootstrap.php"` with `bootstrap="vendor/autoload.php"`, and then in your `Pest.php` file just add `require bootstrap.php` after opening the PHP tag.

That should solve the issues

**Reply by davideprevosto** [reply]
@dingo-d thank you for the reminder, unfortunately warnings are still there.

**Reply by dingo-d** [reply]
Okay, so I've tested my plugin a bit, and this, unfortunately, won't work. At least not for the integration tests for the time being with Pest v2.

The problem is that I didn't update the config correctly for using the correct Integration class (tbh not sure why this worked for me, but I reckon it didn't really work properly).

Once I've made the changes needed, the WordPress test suite started failing, and I forgot that it isn't compatible with PHPUnit 10. I've [made a trac ticket](https://core.trac.wordpress.org/ticket/59486) about updating it to v10, but I wouldn't hold my breath about when it will be made compatible.

**Reply by davideprevosto** [reply]
I read your ticket trac, but I am not so confident that they will follow up in a reasonable time.

**Reply by dingo-d** [reply]
Oh, knowing WordPress, if somebody isn't paid to work on this full-time, this will take a year or two...
